core/trino-spi/src/main/java/io/trino/spi/type/TypeOperators.java

```diff
@@ -69,7 +69,15 @@
     public TypeOperators()
     {
         ConcurrentHashMap<Object, Object> cache = new ConcurrentHashMap<>();
-        this.cache = (operatorConvention, supplier) -> cache.computeIfAbsent(operatorConvention, key -> supplier.get());
+        // optimize the read lock caused by computeIfAbsent
+        // this.cache = (operatorConvention, supplier) -> cache.computeIfAbsent(operatorConvention, key -> supplier.get());
+        this.cache = (operatorConvention, supplier) -> {
+            Object value = cache.get(operatorConvention);
+            if (value == null) {
+                value = cache.computeIfAbsent(operatorConvention, key -> supplier.get());
```

Comment:
I'm surprised this makes a difference.

Reply:
During large-scale data testing, we found that this code has performance problems. When modifying, we also refer to the transformation idea of computeifabsent method of mybatis. The following is the modification of mybatis:

https://github.com/mybatis/mybatis-3/pull/2223

Reply:
The linked JDK bug was fixed in JDK 9. Trino requires JDK 11 at-least. Is there some way to reproduce your test to observe the locking?

Reply:
My microbenchmark was with JDK 11.